smart-contract-extensions/contracts/KeyPurchaser.sol

```diff
@@ -0,0 +1,263 @@
+pragma solidity 0.5.17;
+
+import 'hardlydifficult-ethereum-contracts/contracts/lifecycle/Stoppable.sol';
+import '@openzeppelin/upgrades/contracts/Initializable.sol';
+import '@openzeppelin/contracts/utils/Address.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/SafeERC20.sol';
+import '@openzeppelin/contracts/math/SafeMath.sol';
+import 'unlock-abi-7/IPublicLockV7.sol';
+import './mixins/LockRoles.sol';
+
+/**
+ * @notice Purchase a key priced in any ERC-20 token - either once or as a regular subscription.
+ * This allows the user to purchase or subscribe to a key with 1 tx (`approve`)
+ * or if the token supports it, with 1 signed message (`permit`).
+ *
+ * The user can remove the ERC-20 approval to cancel anytime.
+ *
+ * Risk: if the user transfers or cancels the key, they would naturally expect that also cancels
+ * the subscription but it does not. This should be handled by the frontend.
+ */
+contract KeyPurchaser is Initializable, LockRoles
+{
+  using Address for address payable;
+  using SafeERC20 for IERC20;
+  using SafeMath for uint;
+
+  /**
+   * @notice Emitted when the stop is triggered by a stopper (`account`).
+   */
+  event Stopped(address account);
+
+  // set on initialize and cannot change
+
+  /**
+   * @notice This is the lock for the content users are subscribing to.
+   */
+  IPublicLockV7 public lock;
+
+  /**
+   * @notice The most you will spend on a single key purchase.
+   * @dev This allows the lock owner to increase the key price overtime, up to a point, without
+   * breaking subscriptions.
+   */
+  uint public maxPurchasePrice;
+
+  /**
+   * @notice How close to the end of a subscription someone may renew the purchase.
+   * @dev To ensure that the subscription never expires, we allow this amount of time (maybe 1 day)
+   * before it expires for the renewal to happen.
+   */
+  uint public renewWindow;
+
+  /**
+   * @notice The earliest a renewal may be purchased after the previous purchase was done.
+   * @dev This typically would not apply, but helps to avoid potential abuse or confusion
+   * around cancel and transfer scenarios.
+   */
+  uint public renewMinFrequency;
```

Comment:
I can live with it as is, but the naming here is a bit confusing. It makes me think this is related to how often the key can be renewed. Something like `earliestRenewal` might have clearer semantics.

Reply:
Cancel that. After reading the tests I am clear on how this works/what it means now.